File: apps/mobile/lib/src/helpers/same_day_booking_cutoff.dart
/// When a class running today stops accepting new bookings.
///
/// Staffing for the day is settled in the morning, so from 9am a class running
/// that same day is closed to new bookings (MOB-48). Applies to one-off
/// bookings — by card or by lesson token — and to moving into a class for a
/// single week. It does not apply to admins, who add students to today's
/// classes as the sanctioned way of saying yes to a family who rang up.
///
/// The server enforces the same rule and is the authority. This exists so the
/// option reads as unavailable with a reason, rather than looking bookable and
/// failing after the parent has committed to it.
///
/// **Device time, deliberately.** Like every other date in this app, the
/// comparison is made in the device's timezone rather than Sydney's, so a
/// phone set to another zone will draw the line at its own 9am. That is
/// consistent with the rest of the app rather than a new inconsistency, and
/// MOB-47 — which gives the app real timezone handling — corrects this along
/// with everything else. Until then the server is what makes the rule true:
/// it computes 9am in Sydney and refuses the booking whatever the phone
/// believed.
library;

/// The hour, in the centre's day, from which today's classes are closed.
const int sameDayBookingCutoffHour = 9;

/// The short form, for a class row's subtitle where the full sentence would
/// crowd out the tutor names beside it.
const String sameDayBookingCutoffNote = 'Bookings for today have closed';

/// Shown in place of the option's description once the cutoff has passed.
const String sameDayBookingCutoffHint =
    'Bookings for a class running today close at 9am. Please call us if you '
    'need a place today.';

/// Whether [sessionStartsAt] is closed to new bookings as at [now].
///
/// Closed means the session falls on today's date and it is 9am or later.
/// A session on any other day is not this rule's business, including one that
/// has already run — that is the caller's existing already-started check, and
/// answering `true` here would only confuse the reason shown to the parent.
bool sameDayBookingClosed({
  required DateTime now,
  required DateTime sessionStartsAt,
}) {
  final localNow = now.toLocal();
  final localStart = sessionStartsAt.toLocal();

  final isSameDay = localNow.year == localStart.year &&
      localNow.month == localStart.month &&
      localNow.day == localStart.day;
  if (!isSameDay) return false;

  return localNow.hour >= sameDayBookingCutoffHour;
}

File: backend/firebase/functions/src/attendance/sameDayBookingCutoff.js
"use strict";

/**
 * When a class running today stops accepting new bookings.
 *
 * Staffing for the day is settled in the morning, so a booking that arrives
 * mid-afternoon lands after the roster it affects has been decided. From 9am
 * Sydney a class running that same day is closed to new bookings (MOB-48).
 *
 * Deliberately a calendar-day rule rather than a notice period: "book by 9am
 * for today" is one sentence a parent can hold, where "at least six hours
 * before" has to be recomputed per class.
 *
 * Shared so the payment path, the token path and the swap path close at the
 * same moment. Three copies of "is it past 9am" would eventually disagree,
 * and disagreeing means taking money for a booking another path refuses.
 */

const { DateTime } = require("luxon");

const SYDNEY_ZONE = "Australia/Sydney";

/** Sydney hour from which today's classes are closed. */
const SAME_DAY_CUTOFF_HOUR = 9;

/**
 * What to tell a parent who is too late. One wording, so the app, the payment
 * error and the enrolment error do not each invent their own.
 */
const SAME_DAY_CUTOFF_MESSAGE =
  "Bookings for a class running today close at 9am. Please call us if you need a place today.";

/**
 * Whether [sessionStartsAt] is closed to new bookings as at [now].
 *
 * Both are instants; the zone is applied here rather than by the caller, so a
 * caller cannot get the answer wrong by asking in its own timezone. An
 * attendance document's `date` is the session's start instant, which is what
 * this expects.
 *
 * Closed means: the session falls on today's Sydney date, and it is 9am or
 * later in Sydney. A session on any other date is not this rule's business —
 * including one that has already run, which the capacity and
 * already-started checks handle.
 *
 * Unparseable input is treated as open. A missing or malformed attendance
 * date should not silently refuse a booking that would otherwise be fine; the
 * same direction `sessionHasStarted` leans, and the callers that genuinely
 * require the document to exist check that themselves.
 */
function sameDayBookingClosed({
  sessionStartsAt,
  now = new Date(),
  cutoffHour = SAME_DAY_CUTOFF_HOUR,
  zone = SYDNEY_ZONE,
}) {
  const session = toZoned(sessionStartsAt, zone);
  const current = toZoned(now, zone);
  if (!session || !current) return false;

  if (!session.hasSame(current, "day")) return false;

  return current.hour >= cutoffHour;
}

function toZoned(value, zone) {
  const date =
    value instanceof Date
      ? value
      : typeof value?.toDate === "function"
        ? value.toDate()
        : null;
  if (!date || Number.isNaN(date.getTime())) return null;

  const zoned = DateTime.fromJSDate(date, { zone });
  return zoned.isValid ? zoned : null;
}

module.exports = {
  SAME_DAY_CUTOFF_HOUR,
  SAME_DAY_CUTOFF_MESSAGE,
  SYDNEY_ZONE,
  sameDayBookingClosed,
};
